pkg/agent/controller/traceflow/packetin.go

```diff
@@ -150,16 +160,20 @@ func (c *Controller) parsePacketIn(pktIn *ofctrl.PacketIn) (*crdv1alpha1.Tracefl
 		obs = append(obs, *ob)
 	}
 
-	// Collect Service DNAT.
+	// Collect Service DNAT and SNAT.
 	if !tfState.receiverOnly {
+		ob := new(crdv1alpha1.Observation)
 		if isValidCtNw(ctNwDst) && ipDst != ctNwDst {
-			ob := &crdv1alpha1.Observation{
+			ob = &crdv1alpha1.Observation{
 				Component:       crdv1alpha1.ComponentLB,
 				Action:          crdv1alpha1.ActionForwarded,
 				TranslatedDstIP: ipDst,
 			}
-			obs = append(obs, *ob)
 		}
+		if isValidCtNw(ctNwSrc) && ipSrc != ctNwSrc {
+			ob.TranslatedSrcIP = ipSrc
+		}
```

Comment:
maybe this block should be moved inside of the `if isValidCtNw(ctNwDst) && ipDst != ctNwDst {...}` block

the current code is a bit awkward; it's hard to tell which combinations are valid (I assume it's not possible to have SNAT without DNAT here) and there is no comment about it.

Reply:
@antoninbas do you think it's ok to add a comment like below after move the code?
```
		if isValidCtNw(ctNwDst) && ipDst != ctNwDst {
			ob := &crdv1alpha1.Observation{
				Component:       crdv1alpha1.ComponentLB,
				Action:          crdv1alpha1.ActionForwarded,
				TranslatedDstIP: ipDst,
			}
			// Service SNAT will only happen after DNAT
			if isValidCtNw(ctNwSrc) && ipSrc != ctNwSrc {
				ob.TranslatedSrcIP = ipSrc
			}
			obs = append(obs, *ob)
		}
```

Reply:
Maybe "SNAT may only happen in case of DNAT". And does this only happen for Service hairpinning by the way (I think that's what I observed when I opened the Github issue)? If yes, maybe that's something we could mention in the comment.

Reply:
I checked the pipeline doc, seems there is a [SNAT table](https://github.com/antrea-io/antrea/blob/main/docs/design/ovs-pipeline.md#snattable-71) for egress, so I suppose it's not only for hairpin case?

Reply:
It's a different scenario altogether though, and would not be included in the LB Component observation.

It's orthogonal to this PR, but I don't think we really have support for Egress traffic in Traceflow. I think that we will generate the following observations:
* when Egressing from the same Node, the Forwarding component will report the ForwardedOutOfOverlay action
* when Egressing from a remote Node, the Forwarding component will report the Forwarded action, with the Egress SNAT IP as the tunnel destination IP

@jianjuns and @tnqn can probably give a more definitive answer. It may be nice to have better support for Egress with Traceflow (e.g. report the selected Egress policy), but that should be done as part of a separate issue / PR.

Reply:
OK, if it's a different scenario,maybe I should add a comment like below?
```
// SNAT may only happen in case of DNAT, and is only for hairpin packets at the moment.
```

Reply:
Antonin's description about Egress is correct.